index.d.ts

```diff
@@ -1,47 +1,47 @@
 // Project: Immutability helper
-// TypeScript Version: 2.2
+// TypeScript Version: 2.9
 
 export default update
 
-declare function update<T>(
-  data: ReadonlyArray<T>,
-  query: ArrayOperators<T>,
-): ReadonlyArray<T>
-
-declare function update<T>(
-  data: ReadonlySet<T>,
-  query: SetOperators<T>,
-): ReadonlySet<T>
-
-declare function update<K, V>(
-  data: ReadonlyMap<K, V>,
-  query: MapOperators<K, V>,
-): ReadonlyMap<K, V>
-
-declare function update<T>(data: T, query: Query<T>): T
-
-type Tree<T> = {[K in keyof T]?: Query<T[K]>}
-export type Query<T> =
-  | Tree<T>
-  | ObjectOperators<T>
-  | ArrayOperators<any>
-  | SetOperators<any>
-
-type ObjectOperators<T> =
-  | {$set: any}
-  | {$toggle: Array<keyof T | number>}
-  | {$unset: Array<keyof T | number>}
-  | {$merge: Partial<T>}
-  | {$apply: (old: T) => T}
-  | ((old: T) => any)
-type ArrayOperators<T> =
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+declare function update<T, C extends object = never>(target: T, spec: Spec<T, C>): T
```

Comment:
`C` is the custom commands object. You are required to explicitly type-parameterize this method in order to gain access to your commands.

I could move `C` as the first type argument to make such parameterization more ergonomic, but potentially more confusing at first, as having the object being operated on as the first parameter seems like a better ordering. [Named assignment type parameters](https://github.com/Microsoft/TypeScript/pull/23696) is also coming down the pipe, so this issue will be moot in 3.0 anyway.

Reply:
Will the users have to always pass the second argument, even if they don't use custom commands?